File: hlda_LandSatData_run.py
import torch
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

from hlda_sldr import hlda_sldr # Custom HLDA module

# --- Manual LDA Classifier ---
def manual_lda_classifier(X_train, Y_train, X_test):
    classes = torch.unique(Y_train)
    d = X_train.shape[1]
    pooled_cov = torch.zeros((d, d), dtype=X_train.dtype)
    means, priors = [], []
    for c in classes:
        Xc = X_train[Y_train == c]
        priors.append(Xc.shape[0] / X_train.shape[0])
        means.append(Xc.mean(dim=0))
        pooled_cov += torch.cov(Xc.T) * (Xc.shape[0] - 1)
    pooled_cov /= (X_train.shape[0] - len(classes))
    inv_cov = torch.linalg.inv(pooled_cov)
    
    # Compute LDA decision scores
    scores = []
    for i, mu in enumerate(means):
        mu = mu.view(-1, 1)
        score = X_test @ inv_cov @ mu - 0.5 * mu.T @ inv_cov @ mu + np.log(priors[i])
        scores.append(score.squeeze())
    scores = torch.stack(scores, dim=1)
    preds = classes[scores.argmax(dim=1)]
    return preds

# --- Manual QDA Classifier ---
def manual_qda_classifier(X_train, Y_train, X_test):
    classes = torch.unique(Y_train)
    scores = []
    priors, means, covs = [], [], []

    # Compute class means, priors, and pooled covariance
    for c in classes:
        Xc = X_train[Y_train == c]
        priors.append(Xc.shape[0] / X_train.shape[0])
        means.append(Xc.mean(dim=0))
        covs.append(torch.cov(Xc.T))
    
    # QDA scoring for each test sample
    for i in range(len(classes)):
        mu = means[i]
        cov = covs[i]
        inv_cov = torch.linalg.inv(cov)
        log_det = torch.logdet(cov)
        s = []
        for x in X_test:
            diff = x - mu
            s.append(-0.5 * diff @ inv_cov @ diff - 0.5 * log_det + np.log(priors[i]))
        scores.append(torch.tensor(s))
    scores = torch.stack(scores, dim=1)
    preds = classes[scores.argmax(dim=1)]
    return preds

# --- Plotting ---
def plot_scatter_plain(Z, labels, title_str):
    plt.figure()
    unique_labels = torch.unique(labels)
    colors = [
        [0.8, 0.1, 0.1],   # Class 1 - red
        [0.1, 0.8, 0.1],   # Class 2 - green
        [0.1, 0.1, 0.8],   # Class 3 - blue
        [0.8, 0.8, 0.1],   # Class 4 - yellow/orange
        [0.8, 0.1, 0.8],   # Class 5 - purple
        [0.1, 0.8, 0.8],   # Class 6 - cyan
        [0.5, 0.5, 0.5],   # Class 7 - gray
    ]
    for i, label in enumerate(unique_labels):
        idx = labels == label
        color = colors[i % len(colors)]
        plt.scatter(Z[idx, 0], Z[idx, 1], label=f'Class {int(label)}',
                    color=color, s=35)
    
    # Axis settings
    plt.xlim([-6, 11])
    plt.ylim([-9, 10])
    plt.xticks(np.arange(-6, 12, 2))
    plt.yticks(np.arange(-9, 11, 2))
    plt.title(title_str)
    plt.xlabel("HLDA Dimension 1"); plt.ylabel("HLDA Dimension 2")
    plt.legend(loc='best', fontsize=8)
    plt.grid(True)
    plt.tight_layout()

# ==================
# MAIN EXPERIMENT
# ==================

# Load LandSat data
X_trn = np.loadtxt("sat.trn")
X_tst = np.loadtxt("sat.tst")

X_raw = np.vstack((X_trn[:, :36], X_tst[:, :36]))
Y = np.hstack((X_trn[:, 36], X_tst[:, 36])).astype(int)

# Normalize
scaler = StandardScaler()
X_raw = scaler.fit_transform(X_raw)

X_raw = torch.tensor(X_raw, dtype=torch.float64)
Y = torch.tensor(Y, dtype=torch.int64)

n_runs = 100
target_dim = 3
linear_errors = []
quadratic_errors = []

Z_train_final = Z_test_final = Z_all = None
Y_train_final = Y_test_final = None
mu_last = coeff_last = W_last = None

for run in range(n_runs):
    X_train, X_test, Y_train, Y_test = train_test_split(
        X_raw, Y, test_size=0.1, stratify=Y, random_state=run
    )

    mu_train = X_train.mean(dim=0, keepdim=True)
    X_train_centered = X_train - mu_train
    X_test_centered = X_test - mu_train

    U, S, Vh = torch.linalg.svd(X_train_centered, full_matrices=False)
    latent = S**2 / (X_train_centered.shape[0] - 1)

    sorted_vals, sort_idx = torch.sort(latent, descending=True)
    sorted_vecs = Vh.T[:, sort_idx]

    #sorted_vecs = -sorted_vecs

    if run == 0:
        # Log PCA statistics
        print("PCA Eigendecomposition:")
        print("First 3 eigenvalues:", sorted_vals[:3].numpy())
        print("First eigenvector:", sorted_vecs[:, 0].numpy())
        np.savetxt("pca_first3_eigenvalues_LandSat.csv", sorted_vals[:3].numpy(), delimiter=",")
        np.savetxt("pca_first_eigenvector_LandSat.csv", sorted_vecs[:, 0].numpy().reshape(-1, 1), delimiter=",")
    
    # Select dimensions above numerical threshold
    valid_dims = latent > (1e-6 * latent.sum())
    V = Vh.T[:, valid_dims]
    
    # Project to PCA space
    X_train_pca = X_train_centered @ V
    X_test_pca = X_test_centered @ V
    
    # HLDA Projection
    para, Z_train = hlda_sldr(X_train_pca, Y_train, target_dim)
    Z_test = X_test_pca @ para['W'] - para['mb'] @ para['W']

    if run == 0:
        print("Checking HLDA results:")
        print("HLDA data dimensions:", Z_train.shape)
        print("First 3 data points after HLDA:\n", Z_train[:3].numpy())
        np.savetxt("hlda_first3_data_points_LandSat.csv", Z_train[:3].numpy(), delimiter=",")
    
    # Save final transformation
    if run == n_runs - 1:
        mu_last = mu_train
        coeff_last = V
        W_last = para['W']
        Z_train_final = Z_train
        Z_test_final = Z_test
        Y_train_final = Y_train
        Y_test_final = Y_test
    
    # Classification and error computation
    pred_linear = manual_lda_classifier(Z_train, Y_train, Z_test)
    pred_quad = manual_qda_classifier(Z_train, Y_train, Z_test)

    linear_errors.append((pred_linear != Y_test).float().mean().item())
    quadratic_errors.append((pred_quad != Y_test).float().mean().item())

# === Final full projection
X_centered_full = X_raw - mu_last
X_pca_full = X_centered_full @ coeff_last
Z_all = X_pca_full @ W_last

# === Plots
plot_scatter_plain(Z_all, Y, "Full HLDA Projection (LandSat Dataset)")
plot_scatter_plain(Z_train_final, Y_train_final, "HLDA Projection - Training Set")
plot_scatter_plain(Z_test_final, Y_test_final, "HLDA Projection - Test Set")

# === MCE Plot
plt.figure()
plt.bar(['Linear MCE', 'Quadratic MCE'], [np.mean(linear_errors), np.mean(quadratic_errors)])
plt.ylabel("Classification Error")
plt.title("Mean Classification Error (100 Runs)")
plt.grid(True)
plt.show()

print(f"Average Linear Classification Error: {np.mean(linear_errors):.2f}")
print(f"Average Quadratic Classification Error: {np.mean(quadratic_errors):.2f}")

File: hlda_LandSatData_run.m
% Clean start
close all; clc; clear;

% Load and preprocess LandSat data
train_data = load('sat.trn');
test_data = load('sat.tst');

% Concatenate all data into one array
X_raw = [train_data(:,1:36); test_data(:,1:36)];
Y = [train_data(:,37); test_data(:,37)];

% Normalize entire dataset (Z-score)
X_raw = (X_raw - mean(X_raw)) ./ std(X_raw);

% Preallocate errors
n_runs = 100;
linear_errors = zeros(n_runs,1);
quadratic_errors = zeros(n_runs,1);

target_dim = 3; % HLDA output dimensionality

% Final run storage
Z_train_final = [];
Z_test_final = [];
Y_train_final = [];
Y_test_final = [];

% for full dataset projection (captured after last run)
mu_last = [];
coeff_last = [];
W_last = [];

for run = 1:n_runs
    % Manual 90:10 split
    n = size(X_raw, 1);
    idx = randperm(n);
    n_train = round(0.9 * n);
    train_idx = idx(1:n_train);
    test_idx = idx(n_train+1:end);

    X_train = X_raw(train_idx, :);
    Y_train = Y(train_idx);
    X_test = X_raw(test_idx, :);
    Y_test = Y(test_idx);

    % Center training data
    mu_train = mean(X_train,1);
    X_train_centered = X_train - mu_train;
    X_test_centered = X_test - mu_train;

    % PCA on training data
    [~, S, V] = svd(X_train_centered, 'econ');
    latent = diag(S).^2 / (size(X_train_centered,1) - 1);
    PCA_threshold = 1e-6 * sum(latent);
    valid_dims = find(latent > PCA_threshold);
    coeff = V(:, valid_dims);

    % Project via PCA
    X_train_pca = X_train_centered * coeff;
    X_test_pca = X_test_centered * coeff;
    
    if run == 1
        % Sort latent values and eigenvectors for display
        [sorted_eig_values, sort_idx] = sort(latent, 'descend');
        sorted_eig_vectors = V(:, sort_idx);

        % Print PCA results
        disp('PCA Eigendecomposition:');
        disp('First 3 eigenvalues:');
        disp(sorted_eig_values(1:3)');
        writematrix(sorted_eig_values(1:3), 'pca_first3_eigenvalues_LandSat.csv');
        disp('First eigenvector:');
        disp(sorted_eig_vectors(:, 1)');
        writematrix(sorted_eig_vectors(:, 1), 'pca_first_eigenvector_LandSat.csv');
    end

    % HLDA
    [para, Z_train] = hlda_sldr(X_train_pca, Y_train, target_dim);
    Z_test = X_test_pca * para.W;
    
    if run == 1 
        disp('Checking HLDA results:');
        disp(['HLDA data dimensions: ', num2str(size(Z_train))]);
        disp('First 3 data points after HLDA:');
        disp(Z_train(1:3,:));
    end

    % Store transform from last run
    if run == n_runs
        mu_last = mu_train;
        coeff_last = coeff;
        W_last = para.W;

        Z_train_final = Z_train;
        Z_test_final = Z_test;
        Y_train_final = Y_train;
        Y_test_final = Y_test;
    end

    writematrix(Z_train(1:3,:), 'hlda_first3_data_points_LandSat.csv');


    % Classification
    pred_linear = manual_lda_classifier(Z_train, Y_train, Z_test);
    pred_quad = manual_qda_classifier(Z_train, Y_train, Z_test);

    linear_errors(run) = mean(pred_linear ~= Y_test);
    quadratic_errors(run) = mean(pred_quad ~= Y_test);
end

% ==== Final Full Dataset Projection ====
X_centered_full = X_raw - mu_last;
X_pca_full = X_centered_full * coeff_last;
Z_all = X_pca_full * W_last;

% ==== PLOTS ====
plot_scatter_plain(Z_all, Y, 'Full HLDA Projection (LandSat Dataset)');
legend({'Class 1', 'Class 2', 'Class 3', 'Class 4', 'Class 5', 'Class 6'}, 'Location', 'best');
plot_scatter_plain(Z_train_final, Y_train_final, 'HLDA Projection - Training Set');
legend({'Class 1', 'Class 2', 'Class 3', 'Class 4', 'Class 5', 'Class 6'}, 'Location', 'best');
plot_scatter_plain(Z_test_final, Y_test_final, 'HLDA Projection - Test Set');
legend({'Class 1', 'Class 2', 'Class 3', 'Class 4', 'Class 5', 'Class 6'}, 'Location', 'best');

% MCE plot
figure;
bar([mean(linear_errors), mean(quadratic_errors)]);
set(gca, 'XTickLabel', {'Linear MCE', 'Quadratic MCE'});
ylabel('Classification Error');
title('Mean Classification Error (100 Runs)');
grid on;

% Output
fprintf('Average Linear Classification Error: %.2f\n', mean(linear_errors));
fprintf('Average Quadratic Classification Error: %.2f\n', mean(quadratic_errors));

% ==== Dependencies ====

function preds = manual_lda_classifier(X_train, Y_train, X_test)
    % Implements Linear Discriminant Classifier manually
    classes = unique(Y_train);
    num_classes = numel(classes);
    n_features = size(X_train, 2);
    pooled_cov = zeros(n_features);
    means = zeros(num_classes, n_features);
    priors = zeros(num_classes, 1);
    for i = 1:num_classes
        Xi = X_train(Y_train == classes(i), :);
        priors(i) = size(Xi,1) / size(X_train,1);
        means(i,:) = mean(Xi, 1);
        pooled_cov = pooled_cov + cov(Xi, 1) * (size(Xi,1) - 1);
    end
    pooled_cov = pooled_cov / (size(X_train,1) - num_classes);
    Sigma_inv = inv(pooled_cov);
    scores = zeros(size(X_test,1), num_classes);
    for i = 1:num_classes
        mu = means(i,:)';
        scores(:,i) = X_test * Sigma_inv * mu - 0.5 * mu' * Sigma_inv * mu + log(priors(i));
    end
    [~, idx] = max(scores, [], 2);
    preds = classes(idx);
end

function preds = manual_qda_classifier(X_train, Y_train, X_test)
    % Implements Quadratic Discriminant Classifier manually
    classes = unique(Y_train);
    num_classes = numel(classes);
    n_features = size(X_train, 2);
    means = zeros(num_classes, n_features);
    priors = zeros(num_classes, 1);
    covs = cell(num_classes, 1);
    for i = 1:num_classes
        Xi = X_train(Y_train == classes(i), :);
        priors(i) = size(Xi,1) / size(X_train,1);
        means(i,:) = mean(Xi, 1);
        covs{i} = cov(Xi, 1);
    end
    scores = zeros(size(X_test,1), num_classes);
    for i = 1:num_classes
        mu = means(i,:)';
        Sigma = covs{i};
        Sigma_inv = inv(Sigma);
        log_det = log(det(Sigma));
        for j = 1:size(X_test,1)
            x = X_test(j,:)';
            scores(j,i) = -0.5 * (x - mu)' * Sigma_inv * (x - mu) ...
                          - 0.5 * log_det + log(priors(i));
        end
    end
    [~, idx] = max(scores, [], 2);
    preds = classes(idx);
end

% Scatter plot of 2D projection with class coloring
function plot_scatter_plain(Z, labels, plot_title)
    figure; hold on;
    uniq = unique(labels);
    num_classes = length(uniq);
    colors = [
        0.8 0.1 0.1;
        0.1 0.8 0.1;
        0.1 0.1 0.8;
        0.8 0.8 0.1;
        0.8 0.1 0.8;
        0.1 0.8 0.8;
        0.5 0.5 0.5;
        0.2 0.6 0.2;
        0.6 0.2 0.6
    ];
    for i = 1:num_classes
        idx = labels == uniq(i);
        c = colors(mod(i-1, size(colors,1)) + 1, :);
        plot(Z(idx,1), Z(idx,2), 'o', ...
            'MarkerSize', 6, ...
            'MarkerEdgeColor', c, ...
            'MarkerFaceColor', c);
    end
    title(plot_title);
    xlabel('HLDA Dimension 1'); ylabel('HLDA Dimension 2');
    box on; grid on;
end
